### PR TITLE
Name auctions

### PR DESCRIPTION
**Pull request by sennui**
This is the write up of proposed changes during Berlin meetup.
It covers:
- bidding/auction mechanis
- aet namespace introduction

### CODE REVIEW

File: AENS.md [hunk 1]
```diff
@@ -42,8 +42,7 @@ or `привет.test` are part of the same `.test` namespace.
 
 ## Governance
 
-This first draft of the AENS is not going to have any governance mechanism but
-will only allow registrations under a single namespace: `.test`.
+Lima hard fork is introducing final `.aet` registration namespace. It also retires and purges `.test` namespace.
```
Comment (knarz) [inline AENS.md]: Why will `.test` be removed?

File: AENS.md [hunk 2]
```diff
@@ -66,19 +65,18 @@ could allow us an easier path for future update to the fee structures.
...
 
 To prevent the entry from expiring a user can, at any point before reaching
 the expiration date, update the entry, which pushes the expiration date
-into the future by the time delta of the registration time and the update
-time, e.g. if a user posts an update one week after claiming their name,
-the expiration date gets pushed one week further into the future.
-The main motivation of this expiration date is to prevent situations where
-the private keys, which control the name, are lost making the name unusable
-as well.
+into the future by at most the maximum rent time e.g. if a user posts an update one week after claiming their name, the expiration date gets pushed for a full term from the previous expiration date. Follow up updates will have no effect, until we enter the new rent period. We also introduce rent payment for each update. The rent is locked in lock account and the rent fee is computed based on name length by function defined in governance.
```
Comment (knarz) [inline AENS.md]: So this means that I can renew my name entry only after it entered the expiration period? Why?

File: AENS.md [hunk 3]
```diff
@@ -66,19 +65,18 @@ could allow us an easier path for future update to the fee structures.
 
 Every entry in the `.test` namespace pays the same amount of fees.
 
+The new mechanism planted in Lima hard fork introduces auctions.
+We propose very simplistic scheme. The claim action becomes claim attempt. It can be followed by another claim from an account different than one set in preclaim for given name. There is finite amount of time when the follow up claim is allowed. This time is expressed in height delta computed from function of length of the name. In practice it means that the shorter the name, the longer we wait for another claim attempt. This should work as bidding. Furthermore, the initial fee is also a value of the function of a name length. Each next bid has to be higher by `X` tokens. Finally, we set a protocol protected length of name that is not subject to bidding and will be claimed immediately after confirming claim transaction. All, functions, base fee, free length value and price progression may be subject to changes with governance mechanism. Non-bidding path of the name claim is purposed for development or testing. Initial non-bidding length will be 32.
```
Comment (knarz) [inline AENS.md]: This is really hard to follow. And please format it to 80 characters per line

File: AENS.md (hunk 1 above)
Comment (sennui) [reply AENS.md]: That was the proposal to GC database

File: AENS.md (hunk 2 above)
Comment (sennui) [reply AENS.md]: No, probably I need to re-phrase it.

Intention is following:
1. be similar to DNS re domain management
2. be able to renew at most for the next full term. No matter if you are at the beginning or the end of current term. Basically, the expiration update window is extended to full rent term. Also it limits rent time to one full next term. This is to avoid squating. Extending time window is for UX

File: AENS.md (hunk 3 above)
Comment (sennui) [reply AENS.md]: @knarz I reformatted and re-phrased the algo

File: AENS.md [hunk 4]
```diff
@@ -66,19 +65,47 @@ could allow us an easier path for future update to the fee structures.
...
 Every entry in the `.test` namespace pays the same amount of fees.
 
+The new mechanism planted in Lima hard fork introduces auctions.
+
+The claim action becomes claim attempt. For short names claim transaction
+doesn't set ownership of the name. It can be followed by another
+claim from an account different than one set in preclaim for given name.
+Names below 32 characters are considered short.
+
+There is finite amount of time when the follow up claim is allowed.
```
Comment (knarz) [inline AENS.md]: Why is there a waiting time between claim transactions, when anyone is allowed to send them? Or do you mean claim by the same party?

**Comment by sennui** [comment]
@knarz what details should I elaborate?

There will change in claim transaction format to pass a bid value.
Functions expressing valid bid height were intended to return values in days-weeks.

File: AENS.md [hunk 5]
```diff
@@ -66,19 +65,47 @@ could allow us an easier path for future update to the fee structures.
...
 the expiration date, update the entry, which pushes the expiration date
-into the future by the time delta of the registration time and the update
-time, e.g. if a user posts an update one week after claiming their name,
-the expiration date gets pushed one week further into the future.
-The main motivation of this expiration date is to prevent situations where
-the private keys, which control the name, are lost making the name unusable
-as well.
+into the future by at most the maximum rent time e.g. if a user posts an update
+one week after claiming their name, the expiration date gets pushed for a full
+term from the previous expiration date. Follow up updates will have no effect,
```
Comment (knarz) [inline AENS.md]: So, I registered a name at `registration_height = 500` and we have a maximum rent time of `rent_blocks = 1000`, giving me `termination_height = 1500`. If I now post an update at `update_height = 750`, does that mean `termination_height := termination_height + rent_blocks = 2500 ` or `termination_height := termination_height + (update_height-registration_height) + rent_blocks = 1750`

File: AENS.md [hunk 6]
```diff
@@ -66,19 +65,47 @@ could allow us an easier path for future update to the fee structures.
...
+until we enter the new rent period. We also introduce rent payment for each update.
+The rent is locked in lock account and the rent fee is computed based on name
```
Comment (knarz) [inline AENS.md]: Why is the rent locked and not burned?

Comment (sennui) [reply AENS.md]: The lock account is equivalent of burning. It let us check if there are no tokens missing (or leaking). Also there is an option to run noninflationary airdrop (it would not exceed about of burned=locked tokens).

File: AENS.md (hunk 5 above)
Comment (sennui) [reply AENS.md]: 2500

The change seems to improve UX for business and similar use-cases. Risk of lose of the domain is smaller (easier to set alerts to pay the rent, when termination height is "less" random).

Also, we were thinking to change from ~100 days to annual rent (again, to improve UX).

What you think about it?

File: AENS.md (hunk 4 above)
Comment (sennui) [reply AENS.md]: I think my description is still not clear. 

Claim by different parties.

 Scenario:
1. Pre-claim by X
2. Claim by X at H1 ( T = timout(length(name)) and set bid timeout)
3. Claim by Y at H2 and (H2-H1) < T
4. Claim by X at H3 and (H3-H2) <T
5. ... more than T blocks passes
6. name claimed by X at H4 (H4=H3+T)

If in number 4 X bid at H3' and H3' > T then it would be invalid transaction

Comment (sennui) [reply AENS.md]: The idea is to have classic english type auction. So, we need time frame when next bid is valid.

File: AENS.md (hunk 5 above)
Comment (knarz) [reply AENS.md]: I don't see how that is any less random than the current approach. How is it any better at avoiding the name squatting issue? Because you have to pay a year upfront instead of doing it in smaller increments?

With the old approach, I can basically always set a reminder to `termination_height - (rent_blocks/2)` and expand my name registration for another `rent_blocks` while still having an error margin of `rent_blocks/2`. With your approach, if I understand the »Also it limits rent time to one full next term« correctly, then instead of having a `rent_blocks/2` buffer, you could end up with a `rent_blocks*2` buffer? That's a 4x improvement as far as buffer time goes, but I don't see how that would benefit anyone in practical terms, when the rent time is ~1year.

Comment (sennui) [reply AENS.md]: The `rent_blocks*2` is UX improvement.

I see this becomes controversial and I don't have big opinion here, so maybe we pull product stake holders here.

I understand, tho, the need to mimic more DNS, if we are up for popularizing the platform among regular business. People taking time-off, getting sick,... etc. eventually failing to re-new the name. If we have it happening in the same time of the year and we have months of backup it seems safer.

Current solution would work best, if it is offloaded to 3rd party for management or automated infrastructure.

File: AENS.md [hunk 7]
```diff
@@ -56,15 +56,44 @@ It is unclear what a good mechanism for a naming system would look
...
+After Lima `.test` namespace is no longer available to claim.
+Entries in `.aet` namespace are differentiated regarding fees by their length.
+
+The new mechanism planted in Lima hard fork introduces auctions.
+We make auctions parameters depend on name length.
+
+The claim action becomes claim attempt. For short names claim transaction
+doesn't set ownership of the name. It can be followed by another
+claim from an account different than one set in preclaim for given name.
+Names below 32 characters are considered short.
```
Comment (hanssv) [inline AENS.md]:
```suggestion
Names below 13 characters are considered short.
```

File: AENS.md [hunk 8]
```diff
@@ -168,9 +196,12 @@ long, in order to enable interoperability with DNS.
 A registrar controls who is allowed to claim names in their namespace
 and under what circumstances they are allowed to do so.
 
-The only available registrars for this AENS version will be hard-coded
+The only available registrars up to Lima version will be hard-coded
 ones, which own the `.` and `.test` namespaces.
 
+From Lima we will support `.aet` and limit `.test` namespace.
+Lima disallows write APIs to `.test` namespace.
```
Comment (hanssv) [inline AENS.md]: Weird sentence

File: AENS.md [hunk 9]
```diff
@@ -238,6 +271,98 @@ Note that `expire` is not an explicit message that is part
 of the protocol.
 
 
+### Protocol fees and protection times
+
+Here is the function of initial bidding price depending on name length for Lima hardfork.
+
+```
+ ------------- -------------
```
Comment (hanssv) [inline AENS.md]: Needs an update?

Comment (sennui) [reply AENS.md]: @hanssv No, this is synced with PR to full node. Maybe you mean something else?

File: AENS.md (hunk 7 above)
Comment (sennui) [reply AENS.md]: agrrr. Thanks.

File: AENS.md [hunk 10]
```diff
@@ -186,13 +218,15 @@ to commit to a name and after the commitment has been accepted into the
...
 A commitment should be binding, i.e. the claimant cannot change
-the value they commited to withouth changing the actual commitment
+the value they commited to without changing the actual commitment
 and hiding, so that a malicious miner learns nothing about the value
 the claimant has commited until they chose to reveal that value. This
 prevents malicious miners from front running, i.e. upon seeing a
 claim transaction, including their own claim request for the same
 name instead of the original claimant's one.
 
+After the claim transaction auctioning the name is conducted in clear
```
Comment (jur0) [inline AENS.md]: It doesn't make sense to me, maybe a comma missing somewhere?

File: AENS.md [hunk 11]
```diff
@@ -238,6 +272,98 @@ Note that `expire` is not an explicit message that is part
 of the protocol.
 
 
+### Protocol fees and protection times
```
Comment (dincho) [inline AENS.md]: The fee in the table is missing units, as we discussed with @sennui it's 10^14 e.g. 3e14 - 5702887e14 (~570AE)

**Comment by randomshinichi** [comment]
Please update `serializations.md` to reflect the RLP serialization of NameUpdateTx

File: AENS.md [hunk 12]
```diff
@@ -56,15 +56,44 @@ It is unclear what a good mechanism for a naming system would look
 like. If we imagine two actors both being interested in the same name,
```
[the remark this answers is not in the document]
Comment (sennui) [reply AENS.md]: lets stick to minimal change set... This is original question by Sasha to set the stage.

File: AENS.md [hunk 13]
```diff
@@ -56,15 +56,44 @@ It is unclear what a good mechanism for a naming system would look
...
 Every entry in the `.test` namespace pays the same amount of fees.
+After Lima `.test` namespace is no longer available to claim.
+Entries in `.aet` namespace are differentiated regarding fees by their length.
+
+The new mechanism planted in Lima hard fork introduces auctions.
+We make auctions parameters depend on name length.
+
+The claim action becomes claim attempt. For short names claim transaction
+doesn't set ownership of the name. It can be followed by another
+claim from an account different than one set in preclaim for given name.
```
[the remark this answers is not in the document]
Comment (sennui) [reply AENS.md]: rephrased

File: AENS.md [hunk 14]
```diff
@@ -73,12 +102,11 @@ it could be claimed again.
 
 To prevent the entry from expiring a user can, at any point before reaching
```
[the remark this answers is not in the document]
Comment (sennui) [reply AENS.md]: I am not sure about this one. Will ask native English speaker to redact.

File: AENS.md (hunk 10 above)
Comment (sennui) [reply AENS.md]: rephrased

**Comment by sennui** [comment]
It got unreadable after master rebase, here is follow up PR: https://github.com/aeternity/protocol/pull/428

**Comment by sennui** [comment]
Closing to clear the thread, new PR: #428. 
All reviews suggestions were considered.
Some resolved without action, as no longer relevant (e.g. we keep the old "rent" mechanism)

File: AENS.md [hunk 15]
```diff
@@ -73,12 +102,11 @@ it could be claimed again.
...
-into the future by the time delta of the registration time and the update
-time, e.g. if a user posts an update one week after claiming their name,
-the expiration date gets pushed one week further into the future.
-The main motivation of this expiration date is to prevent situations where
-the private keys, which control the name, are lost making the name unusable
-as well.
+into the future by the time delta of the registration time and the update time,
+e.g. if a user posts an update one week after claiming their name, the expiration
+date gets pushed one week further into the future. The main motivation of this
+expiration date is to prevent situations where the private keys, which control
```
[the remark this answers is not in the document]
Comment (sennui) [reply AENS.md]: With current tx API: no. Our curve would support that, tho.
